From fc381c5ef6f87b6eb16a559c0c91b6e30533e615 Mon Sep 17 00:00:00 2001
From: Han Xu <b45815@freescale.com>
Date: Fri, 9 Oct 2015 10:18:20 -0500
Subject: [PATCH 0718/5242] MLK-11719-2: mtd: gpmi: save the bch layout
 setting in debugfs

commit  82f38ebf0e723cc3682e12cc82ef99f6048413e9 from
https://source.codeaurora.org/external/imx/linux-imx.git

save the bch layout setting in debugfs for the upper layer applications,
such as kobs-ng.

Signed-off-by: Han Xu <b45815@freescale.com>
(cherry picked from commit 8a373e796c21f4e9b714039e5f0b7d9388ef5a32)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c |   19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c b/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c
index 8cb2586..98d1462 100644
--- a/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c
+++ b/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c
@@ -22,12 +22,16 @@
 #include <linux/clk.h>
 #include <linux/slab.h>
 #include <linux/pm_runtime.h>
+#include <linux/debugfs.h>
 
 
 #include "gpmi-nand.h"
 #include "gpmi-regs.h"
 #include "bch-regs.h"
 
+/* export the bch geometry to dbgfs */
+static struct debugfs_blob_wrapper dbg_bch_geo;
+
 /* Converts time to clock cycles */
 #define TO_CYCLES(duration, period) DIV_ROUND_UP_ULL(duration, period)
 
@@ -263,6 +267,7 @@ int bch_set_geometry(struct gpmi_nand_data *this)
 	unsigned int page_size;
 	unsigned int gf_len;
 	int ret;
+	struct dentry *dbg_root;
 
 	ret = common_nfc_set_geometry(this);
 	if (ret)
@@ -275,6 +280,20 @@ int bch_set_geometry(struct gpmi_nand_data *this)
 	page_size     = bch_geo->page_size;
 	gf_len        = bch_geo->gf_len;
 
+	dbg_root = debugfs_create_dir("gpmi-nand", NULL);
+	if (!dbg_root) {
+		dev_err(this->dev, "failed to create debug directory\n");
+		return -EINVAL;
+	}
+
+	dbg_bch_geo.data = (void *)bch_geo;
+	dbg_bch_geo.size = sizeof(struct bch_geometry);
+	if (!debugfs_create_blob("bch_geometry", S_IRUGO,
+				dbg_root, &dbg_bch_geo)) {
+		dev_err(this->dev, "failed to create debug bch geometry\n");
+		return -EINVAL;
+	}
+
 	ret = pm_runtime_get_sync(this->dev);
 	if (ret < 0) {
 		dev_err(this->dev, "Failed to enable clock\n");
-- 
1.7.9.5

